Handle signals properly when dropping GVL

murgatroid99 opened the pull request:
This is an improvement on #6472. We now no longer have extra ruby code to handle signals, we no longer spawn an extra thread to handle signals, and it no longer takes up to a second to handle a signal. Instead, we now use a built in facility to temporarily stop plucking when a signal arrives.

murgatroid99 commented:
All of the Ruby tests pass.

jtattermusch commented on src/ruby/ext/grpc/rb_completion_queue.c:
```diff
@@ -158,8 +189,23 @@ grpc_event grpc_rb_completion_queue_pluck_event(VALUE self, VALUE tag,
+
+     The basic reason we need this relatively complicated construction is that
+     we need to re-acquire the GVL when an interrupt comes in, so that the ruby
+     interpeter can do what it needs to do with the interrupt. But we also need
```
typo: interpeter

unfinished sentence

murgatroid99 replied:
Fixed

jtattermusch commented on src/ruby/ext/grpc/rb_completion_queue.c:
```diff
@@ -52,21 +52,47 @@ typedef struct next_call_stack {
   next_call_stack *const next_call = (next_call_stack*)param;
-  next_call->event =
-      grpc_completion_queue_next(next_call->cq, next_call->timeout, NULL);
+  gpr_timespec increment = gpr_time_from_millis(20, GPR_TIMESPAN);
```
do we really have to wake up 50x a second?

murgatroid99 replied:
It's a time period that's short from the point of view of a human, but long in terms of processing time. In the future, we can tune it if it turns out to have a significant performance impact.

jtattermusch replied:
I see the point, but in general I think these kinds of tricks are problematic. It's probably fine with me but I really wish there was a cleaner solution.

murgatroid99 replied:
Optimally, there would be a function to wake/fail an existing call to `pluck`. But we don't have that, and this is the best option I can see.

jtattermusch commented on src/ruby/ext/grpc/rb_completion_queue.c:
```diff
@@ -52,21 +52,47 @@ typedef struct next_call_stack {
+  gpr_timespec increment = gpr_time_from_millis(20, GPR_TIMESPAN);
+  gpr_timespec deadline;
+  do {
+    deadline = gpr_time_add(gpr_now(GPR_CLOCK_REALTIME), increment);
```
in next_call->timeout is in 19msec and our deadline is in 20, we just skip?

murgatroid99 replied:
Fixed. I moved the check to after `pluck` finishes.

murgatroid99 commented:
In addition to those fixes, I also changed my mind again and removed the ruby-level signal handling. The code I have seems to do fine at interrupting `pluck` with an `Interrupt` exception when I press Ctrl+C.

jtattermusch commented:
LGTM.

murgatroid99 commented:
OK, I think this can be merged. There was a problem that I was looking into, but it appears to not be related to this change.
